### PR TITLE
[FLINK-11405][rest]rest api can more exceptions by query parameter

### PR DESCRIPTION
**Pull request by jinglining**
## What is the purpose of the change
This pull request makes rest api can show more exceptions by query parameter.
User could define exception's size  to view by  /jobs/:jobid/exceptions?size=[size]

## Brief change log
- Add JobExceptionsMessageParameters and ExceptionShowSizeParameter for define size in url.
- Update JobExceptionsHandler

## Verifying this change

Add JobExceptionsHandlerTest.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no )
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not documented)

### CODE REVIEW

**Comment by zentol** [comment]
Please provide a description on what exactly you're trying to accomplish in the JIRA.

**Comment by jinglining** [comment]
> Please provide a description on what exactly you're trying to accomplish in the JIRA.

I have updated the JIAR.

**Comment by jinglining** [comment]
@zentol  I have update the JIRA, could you open it?

File: flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/job/JobExceptionsHandler.java [hunk 1]
```diff
@@ -49,15 +50,15 @@
 /**
  * Handler serving the job exceptions.
  */
-public class JobExceptionsHandler extends AbstractExecutionGraphHandler<JobExceptionsInfo, JobMessageParameters> implements JsonArchivist {
+public class JobExceptionsHandler extends AbstractExecutionGraphHandler<JobExceptionsInfo, JobExceptionsMessageParameters> implements JsonArchivist {
 
-	static final int MAX_NUMBER_EXCEPTION_TO_REPORT = 20;
+	private static int maxNumberExceptionToReport = 20;
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/job/JobExceptionsHandler.java]: it's a bad idea to use a mutable field for passing arguments around. Revert this change and add another argument to `createJobExceptionsInfo()` instead and propagate it to `handleRequest()`.

File: flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/job/JobExceptionsHandler.java [hunk 2]
```diff
@@ -71,7 +72,11 @@ public JobExceptionsHandler(
 	}
 
 	@Override
-	protected JobExceptionsInfo handleRequest(HandlerRequest<EmptyRequestBody, JobMessageParameters> request, AccessExecutionGraph executionGraph) {
+	protected JobExceptionsInfo handleRequest(HandlerRequest<EmptyRequestBody, JobExceptionsMessageParameters> request, AccessExecutionGraph executionGraph) {
+		List<Integer> sizes = request.getQueryParameter(ExceptionShowSizeParameter.class);
+		if (sizes != null && sizes.size() == 1) {
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/job/JobExceptionsHandler.java]: the returned list is never null. The size check is overly strict and should just check `sizes.size() > 0`. It is fine to use the first value should the parameter be defined multiple times.

File: flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/job/JobExceptionsHandler.java [hunk 3]
```diff
@@ -71,7 +72,11 @@ public JobExceptionsHandler(
 	}
 
 	@Override
-	protected JobExceptionsInfo handleRequest(HandlerRequest<EmptyRequestBody, JobMessageParameters> request, AccessExecutionGraph executionGraph) {
+	protected JobExceptionsInfo handleRequest(HandlerRequest<EmptyRequestBody, JobExceptionsMessageParameters> request, AccessExecutionGraph executionGraph) {
+		List<Integer> sizes = request.getQueryParameter(ExceptionShowSizeParameter.class);
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/job/JobExceptionsHandler.java]: variable should have a better name.

File: flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/job/ExceptionShowSizeParameter.java [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+	}
+
+	@Override
+	public String convertValueToString(Integer value) {
+		return value.toString();
+	}
+
+	@Override
+	public String getDescription() {
+		return "Show exception's size.";
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/job/ExceptionShowSizeParameter.java]: Parameter descriptions are reasonable standardized. Please have a look at other descriptions and update this one accordingly.

Additionally the description doesn't really make sense; we aren't changing the size of exceptions after all.

Here's a suggestion to save us some time, but it would be good if you would be more diligent in the future:

```suggestion
		return "Integer value that specifies the upper limit of exceptions to return.";
```

File: flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/job/ExceptionShowSizeParameter.java [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+
+import org.apache.flink.runtime.rest.messages.MessageQueryParameter;
+
+/**
+ * @see org.apache.flink.runtime.rest.handler.job.JobExceptionsHandler
+ * show size for JobExceptionsHandler
+ */
+public class ExceptionShowSizeParameter extends MessageQueryParameter<Integer> {
+
+	private static final String QUERY_PARAMETER_NAME = "size";
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/job/ExceptionShowSizeParameter.java]: needs a better name

Comment (jinglining) [reply flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/job/ExceptionShowSizeParameter.java]: How about limit?

File: flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/job/JobExceptionsHandler.java (hunk 2 above)
Comment (jinglining) [reply flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/job/JobExceptionsHandler.java]: As MessageQueryParameter's type is List<X>, sizes  has to define to List<Integer>. It's no good. Maybe we could split MessageQueryParameter to MessageQueryListParameter and MessageQueryObjectParameter?

**Comment by jinglining** [comment]
@flinkbot run travis

File: flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/job/JobExceptionsHandler.java (hunk 2 above)
Comment (zentol) [reply flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/job/JobExceptionsHandler.java]: The fact of the matter is that users can provide multiple value for query parameters. Hence, you have to handle this case somehow.

Again, just use the first value.

File: flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/JobExceptionsInfo.java [hunk 6]
```diff
@@ -81,6 +81,22 @@ public int hashCode() {
 		return Objects.hash(rootException, rootTimestamp, allExceptions, truncated);
 	}
 
+	public String getRootException() {
+		return rootException;
+	}
+
+	public Long getRootTimestamp() {
+		return rootTimestamp;
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/JobExceptionsInfo.java]: getters should be annotated with `@JsonIgnore` to prevent wacky behaviors if the java/json field names differ.

File: flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/job/UpperLimitExceptionParameter.java [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+	}
+
+	@Override
+	public String convertValueToString(Integer value) {
+		return value.toString();
+	}
+
+	@Override
+	public String getDescription() {
+		return "Integer value that specifies the upper limit of exceptions to return.";
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/job/UpperLimitExceptionParameter.java]: this isn't in sync with the documentation? (The description in the docs is better)

File: docs/_includes/generated/rest_v1_dispatcher.html [hunk 8]
```diff
@@ -1893,6 +1893,16 @@
         </ul>
       </td>
     </tr>
+    <tr>
+        <td colspan="2">Query parameters</td>
+    </tr>
+    <tr>
+        <td colspan="2">
+            <ul>
+                <li><code>size</code> (optional): Comma-separated list of integer values that specifie the upper limit of exceptions to return.</li>
```
Comment (zentol) [inline docs/_includes/generated/rest_v1_dispatcher.html]:
```suggestion
                <li><code>size</code> (optional): Comma-separated list of integer values that specifies the upper limit of exceptions to return.</li>
```

File: flink-runtime/src/test/java/org/apache/flink/runtime/rest/handler/job/JobExceptionsHandlerTest.java [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+		final AccessExecutionGraph archivedExecutionGraph = new ArchivedExecutionGraphBuilder()
+			.setTasks(tasks)
+			.build();
+		final int querySize = 10;
+		int finalShowSize = exceptionSize >= querySize ? querySize : exceptionSize;
+		final HandlerRequest<EmptyRequestBody, JobExceptionsMessageParameters> handlerRequest = createRequest(archivedExecutionGraph.getJobID(), querySize);
+
+		final JobExceptionsInfo jobExceptionsInfo = jobExceptionsHandler.handleRequest(handlerRequest, archivedExecutionGraph);
+
+		assert(jobExceptionsInfo.getAllExceptions().size() == finalShowSize);
```
Comment (zentol) [inline flink-runtime/src/test/java/org/apache/flink/runtime/rest/handler/job/JobExceptionsHandlerTest.java]: use assertEquals instead

File: flink-runtime/src/test/java/org/apache/flink/runtime/rest/handler/job/JobExceptionsHandlerTest.java [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+		final int exceptionSize = 50;
+		for (int i = 0; i < exceptionSize; i++) {
+			final JobVertexID jobVertexId = new JobVertexID();
+			tasks.put(jobVertexId, createArchivedExecutionJobVertex(jobVertexId));
+		}
+		final AccessExecutionGraph archivedExecutionGraph = new ArchivedExecutionGraphBuilder()
+			.setTasks(tasks)
+			.build();
+		final int querySize = 10;
+		int finalShowSize = exceptionSize >= querySize ? querySize : exceptionSize;
```
Comment (zentol) [inline flink-runtime/src/test/java/org/apache/flink/runtime/rest/handler/job/JobExceptionsHandlerTest.java]: you should explicitly test these different cases.

File: flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/job/UpperLimitExceptionParameter.java [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.rest.messages.job;
+
+import org.apache.flink.runtime.rest.messages.MessageParameter;
+import org.apache.flink.runtime.rest.messages.MessageQueryParameter;
+
+/**
+ * @see org.apache.flink.runtime.rest.handler.job.JobExceptionsHandler
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/job/UpperLimitExceptionParameter.java]: Move the second line to the top and add a period at the end. `@see` references are commonly put last in the javadocs.

File: flink-runtime/src/test/java/org/apache/flink/runtime/rest/handler/job/JobExceptionsHandlerTest.java [hunk 12]
```diff
@@ -0,0 +1,137 @@
...
+import org.apache.flink.util.TestLogger;
+
+import org.junit.Test;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+
```
Comment (zentol) [inline flink-runtime/src/test/java/org/apache/flink/runtime/rest/handler/job/JobExceptionsHandlerTest.java]: remove empty line

File: flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/job/ExceptionShowSizeParameter.java (hunk 5 above)
Comment (jinglining) [reply flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/job/ExceptionShowSizeParameter.java]: Ok.

File: flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/job/JobExceptionsHandler.java [hunk 13]
```diff
@@ -80,7 +82,7 @@ protected JobExceptionsInfo handleRequest(HandlerRequest<EmptyRequestBody, JobEx
 
 	@Override
 	public Collection<ArchivedJson> archiveJsonWithPath(AccessExecutionGraph graph) throws IOException {
-		ResponseBody json = createJobExceptionsInfo(graph, MAX_NUMBER_EXCEPTION_TO_REPORT);
+		ResponseBody json = createJobExceptionsInfo(graph, Iterables.size(graph.getAllExecutionVertices()));
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/job/JobExceptionsHandler.java]: This change is out-of-scope of this PR and is a bit dangerous since it can significantly change current behavior; for example if you have a massive deployment and all of them fail for some reason (like a class-loading issue).

I'd revert it for the time being.

File: flink-runtime/src/test/java/org/apache/flink/runtime/rest/handler/job/JobExceptionsHandlerTest.java [hunk 14]
```diff
@@ -66,23 +69,30 @@ public void testGetJobExceptionsInfo() throws HandlerRequestException {
...
+			final int querySize = exceptionSize2QuerySize.f1;
+			final AccessExecutionGraph archivedExecutionGraph = createAccessExecutionGraph(exceptionSize);
+			int finalShowSize = exceptionSize >= querySize ? querySize : exceptionSize;
+			final HandlerRequest<EmptyRequestBody, JobExceptionsMessageParameters> handlerRequest = createRequest(archivedExecutionGraph.getJobID(), querySize);
+			final JobExceptionsInfo jobExceptionsInfo = jobExceptionsHandler.handleRequest(handlerRequest, archivedExecutionGraph);
+			assertEquals(jobExceptionsInfo.getAllExceptions().size(), finalShowSize);
+		}
+	}
 
+	private AccessExecutionGraph createAccessExecutionGraph(int taskSize) {
```
Comment (zentol) [inline flink-runtime/src/test/java/org/apache/flink/runtime/rest/handler/job/JobExceptionsHandlerTest.java]: rename `taskSize` to `numTasks`

Can we make this method static, along with `createArchivedExecutionJobVertex` and `createRequest`?

File: flink-runtime/src/test/java/org/apache/flink/runtime/rest/handler/job/JobExceptionsHandlerTest.java [hunk 15]
```diff
@@ -66,23 +69,30 @@ public void testGetJobExceptionsInfo() throws HandlerRequestException {
 			JobExceptionsHeaders.getInstance(),
 			new ExecutionGraphCache(TestingUtils.TIMEOUT(), TestingUtils.TIMEOUT()),
 			TestingUtils.defaultExecutor());
+		List<Tuple2<Integer, Integer>> exceptionSize2QuerySizeList = new ArrayList<>();
+		exceptionSize2QuerySizeList.add(new Tuple2<>(50, 20));
+		exceptionSize2QuerySizeList.add(new Tuple2<>(30, 30));
+		exceptionSize2QuerySizeList.add(new Tuple2<>(10, 20));
+		for (Tuple2<Integer, Integer> exceptionSize2QuerySize: exceptionSize2QuerySizeList) {
```
Comment (zentol) [inline flink-runtime/src/test/java/org/apache/flink/runtime/rest/handler/job/JobExceptionsHandlerTest.java]: I think we can simplify this as we don't need multiple execution graphs, one is sufficient:

```
	...
	final int numExceptions = 30;
	final AccessExecutionGraph archivedExecutionGraph = createAccessExecutionGraph(numExceptions);
	checkExceptionLimit(archivedExecutionGraph, 50, numExceptions);
	checkExceptionLimit(archivedExecutionGraph, numExceptions, numExceptions);
	checkExceptionLimit(archivedExecutionGraph, 10, 10);
}


private static void checkExceptionLimit(AccessExecutionGraph graph, int maxNumExceptions, int numExpectedException) {
	final HandlerRequest<EmptyRequestBody, JobExceptionsMessageParameters> handlerRequest = createRequest(graph.getJobID(), maxNumExceptions);
	final JobExceptionsInfo jobExceptionsInfo = jobExceptionsHandler.handleRequest(handlerRequest, graph);
	assertEquals(jobExceptionsInfo.getAllExceptions().size(), numExpectedException);
}
```

**Comment by jinglining** [comment]
@flinkbot run travis

**Comment by jinglining** [comment]
@flinkbot run travis
